scope_shared/scope/populate/data/validate.py

```diff
@@ -0,0 +1,424 @@
+        # This would also be caught in the next loop,
+        # but is explicitly tested here for clarity in case of failure.
+        assert revisions_current.filter_match(
+            match_values={
+                "_rev": 1
+            }
+        ).is_unique()
+
+        document_previous = None
+        for index, document_current in enumerate(revisions_current.order_by_revision()):
+            # Revisions must have an incrementing sequence of _rev values
+            assert document_current["_rev"] == index + 1
+
+            # Any deletion cannot be the first revision and must be the final revision
+            if "_delete" in document_current:
+                assert "_rev" != 1
+                assert "_rev" == len(revisions_current)
+
+            # Times must be monotonically increasing.
+            if document_previous:
+                assert datetime_from_document(document=document_current) >= datetime_from_document(document=document_previous)
+
+            document_previous = document_current
+
+
+def _validate_patient_collection_strings(
+    *,
+    collection: DocumentSet,
+):
+    """
+    Validate properties of strings in specific documents.
+    """
+
+    activity_documents = collection.filter_match(
+        match_type="activity",
+        match_deleted=False,
+    )
+    value_documents = collection.filter_match(
+        match_type="value",
+        match_deleted=False,
+    )
+
+    # Activity names must already be trimmed
+    for document_current in activity_documents:
+        assert document_current["name"] == document_current["name"].strip()
+
+    # Value names must already be trimmed
+    for document_current in value_documents:
+        assert document_current["name"] == document_current["name"].strip()
+
+
+def _validate_patient_collection_activity(*, collection: DocumentSet,):
+    """
+    Validate additional properties of activity documents.
+    """
+
+    activity_documents = collection.filter_match(
+        match_type="activity",
+    )
+    value_documents = collection.filter_match(
+        match_type="value",
+    )
+
+    for document_current in activity_documents.filter_match(
+        match_deleted=False,
+    ):
+        # If the activity has a value, the referenced value must exist.
+        if "valueId" in document_current:
+            assert value_documents.filter_match(
+                match_deleted=False,
+                match_datetime_at=datetime_from_document(document=document_current),
+                match_values={
+                    "valueId": document_current["valueId"]
+                }
+            ).is_unique()
+
+        # For the time that each activity exists, its name/value tuple must be unique.
+        assert activity_documents.filter_match(
+            match_datetime_at=datetime_from_document(document=document_current),
+            match_deleted=False,
+            match_values={
+                "name": document_current["name"],
```

Comment:
I believe we enforce the activity's name to be unique across all values in the client.

Reply:
Yes that's right, fixed in 43d1853